src/connectors/impls/gbq/writer/sink.rs

```diff
@@ -0,0 +1,660 @@
+        &self.descriptor
+    }
+}
+impl GbqSink {
+    pub async fn new(config: Config) -> Result<Self> {
+        Ok(Self {
+            client: None,
+            write_stream: None,
+            mapping: None,
+            config,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for GbqSink {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        _serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        let client = self
+            .client
+            .as_mut()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                "The client is not connected",
+            ))?;
+        let write_stream =
+            self.write_stream
+                .as_ref()
+                .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                    "The write stream is not available",
+                ))?;
+        let mapping = self
+            .mapping
+            .as_ref()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                "The mapping is not available",
+            ))?;
+
+        let mut serialized_rows = vec![];
+
+        for data in event.value_iter() {
+            serialized_rows.push(mapping.map(data)?);
+        }
+
+        let request = AppendRowsRequest {
+            write_stream: write_stream.name.clone(),
+            offset: None,
+            trace_id: "".to_string(),
+            rows: Some(append_rows_request::Rows::ProtoRows(ProtoData {
+                writer_schema: Some(ProtoSchema {
+                    proto_descriptor: Some(mapping.descriptor().clone()),
+                }),
+                rows: Some(ProtoRows { serialized_rows }),
+            })),
+        };
+
+        let append_response = client
+            .append_rows(stream::iter(vec![request]))
+            .timeout(Duration::from_secs(self.config.request_timeout))
+            .await;
+
+        let append_response = if let Ok(append_response) = append_response {
+            append_response
+        } else {
+            ctx.notifier.connection_lost().await?;
+
+            return Ok(SinkReply::FAIL);
+        };
+
+        if let Ok(x) = append_response?
+            .into_inner()
+            .next()
+            .timeout(Duration::from_secs(self.config.request_timeout))
+            .await
+        {
+            match x {
+                Some(Ok(_)) => Ok(SinkReply::ACK),
+                Some(Err(e)) => {
+                    error!("BigQuery error: {}", e);
+
+                    Ok(SinkReply::FAIL)
+                }
+                None => Ok(SinkReply::NONE),
+            }
+        } else {
+            ctx.notifier.connection_lost().await?;
+
+            Ok(SinkReply::FAIL)
+        }
+    }
+
+    async fn connect(&mut self, _ctx: &SinkContext, _attempt: &Attempt) -> Result<bool> {
+        error!("Connecting to BigQuery");
+        let token = Token::new()?.header_value()?;
```

Comment:
oh, that's a good point, we should be passing the `Token` struct into the interceptor, and calling `header_value` on every request (it checks for expiration and refreshes as needed)

Reply:
You'd need to put the `Token` into the `AuthInterceptor` and for each request call `Token::header_value` this will do the refreshing bits: https://docs.rs/gouth/0.2.1/src/gouth/token.rs.html#24-38

Reply:
Didn't see your comment! :) We were thinking the same!